File: Observer/observer.go
package main

// 代码参考 https://blog.csdn.net/Jeanphorn/article/details/78784197 表示感谢

import (
	"fmt"
)

type Event struct {
	Data string
}

type Observer interface {
	update(event *Event)
}

type Subject interface {
	Attach(observer Observer)
	Detach(observer Observer)
	Notify(event *Event)
}

type ConcreteObserver struct {
	Id int64
}

func (co *ConcreteObserver) update(event *Event) {
	fmt.Println(co.Id, "update: ", event.Data)

}

type ConcreteSubject struct {
	Observers map[Observer]interface{}
}

func (cs *ConcreteSubject) Attach(observer Observer) {
	cs.Observers[observer] = nil
}

func (cs *ConcreteSubject) Detach(observer Observer) {
	delete(cs.Observers, observer)
}

func (cs *ConcreteSubject) Notify(event *Event) {
	for observer, _ := range cs.Observers {
		observer.update(event)
	}
}

func main() {
	cs := ConcreteSubject{
		Observers: make(map[Observer]interface{}),
	}
	cs1 := ConcreteObserver{1}
	cs2 := ConcreteObserver{2}
	cs.Attach(&cs1)
	cs.Attach(&cs2)
	event := Event{"hello"}
	cs.Notify(&event)
	fmt.Println("remove 2")
	cs.Detach(&cs2)
	cs.Notify(&event)
}

File: Observer/observer.py
import abc


class Subject(object):
    def __init__(self):
        self._observers = set()

    def attach(self, observer):
        self._observers.add(observer)

    def detach(self, observer):
        self._observers.remove(observer)

    def notify(self, data):
        for observer in self._observers:
            observer.update(data)


class Observer(object):
    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def update(self, data):
        pass


class Observer1(Observer):

    def update(self, data):
        print(self.__class__.__name__ + ": " + data)


class Observer2(Observer):

    def update(self, data):
        print(self.__class__.__name__ + ": " + data)


if __name__ == '__main__':
    subject = Subject()
    observer1 = Observer1()
    observer2 = Observer2()
    subject.attach(observer1)
    subject.attach(observer2)
    subject.notify("hello")

    print("remove observer2")

    subject.detach(observer2)
    subject.notify("hello")



